pkg/cloudprovider/provider/azure/provider.go

```diff
@@ -0,0 +1,581 @@
+			NetworkProfile: &compute.NetworkProfile{
+				NetworkInterfaces: &[]compute.NetworkInterfaceReference{
+					{
+						ID: iface.ID,
+						NetworkInterfaceReferenceProperties: &compute.NetworkInterfaceReferenceProperties{Primary: to.BoolPtr(true)},
+					},
+				},
+			},
+			OsProfile: &compute.OSProfile{
+				AdminUsername: to.StringPtr(adminUserName),
+				ComputerName:  &machine.Spec.Name,
+				LinuxConfiguration: &compute.LinuxConfiguration{
+					DisablePasswordAuthentication: to.BoolPtr(true),
+					SSH: &compute.SSHConfiguration{
+						PublicKeys: &[]compute.SSHPublicKey{
+							{
+								Path:    to.StringPtr(fmt.Sprintf("/home/%s/.ssh/authorized_keys", adminUserName)),
+								KeyData: &key.PublicKey,
+							},
+						},
+					},
+				},
+				CustomData: to.StringPtr(base64.StdEncoding.EncodeToString([]byte(userdata))),
+			},
+			StorageProfile: &compute.StorageProfile{ImageReference: osRef},
+		},
+		Tags: map[string]*string{machineUIDTag: to.StringPtr(string(machine.UID))},
+	}
+
+	glog.Infof("Creating machine %q", machine.Spec.Name)
+	future, err := vmClient.CreateOrUpdate(context.TODO(), config.ResourceGroup, machine.Spec.Name, vmSpec)
+	if err != nil {
+		return nil, fmt.Errorf("trying to create a VM: %s", err)
+	}
+
+	err = future.WaitForCompletion(context.TODO(), vmClient.Client)
+	if err != nil {
+		return nil, fmt.Errorf("waiting for operation returned: %s", err.Error())
+	}
+
+	vm, err := future.Result(vmClient)
+	if err != nil {
+		return nil, fmt.Errorf("decoding result: %s", err.Error())
+	}
+
+	// get the actual VM object filled in with additional data
+	vm, err = vmClient.Get(context.TODO(), config.ResourceGroup, machine.Spec.Name, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve updated data for VM %q: %s", machine.Spec.Name, err)
+	}
+
+	ipAddresses, err := getVMIPAddresses(context.TODO(), config, &vm)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve IP addresses for VM %q: %s", machine.Spec.Name, err.Error())
+	}
+
+	status, err := getVMStatus(context.TODO(), config, machine.Spec.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve status for VM %q: %s", machine.Spec.Name, err.Error())
+	}
+
+	return &azureVM{vm: &vm, ipAddresses: ipAddresses, status: status}, nil
+}
+
+func (p *provider) Delete(machine *v1alpha1.Machine, instance instance.Instance) error {
+	config, _, err := p.getConfig(machine.Spec.ProviderConfig)
+	if err != nil {
+		return cloudprovidererrors.TerminalError{
+			Reason:  v1alpha1.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	// TODO: Disassociate and remove the resources in reverse order, to prevent
+	// orphaned NICs, IPs and disks in case the operation is interrupted in the middle.
+	glog.Infof("deleting VM %q", machine.Name)
+	if err = deleteVMsByMachineUID(context.TODO(), config, machine.UID); err != nil {
```

Comment:
This wasn't addresses, was it?

Reply:
I don't see how we can recover from a partial error here. In case of a VM not found, this will simply return nil, since we're iterating through the list of all VMs.

Reply:
Right, missed it. This function can be called over and over again, so it indeed recovers correctly